glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,335 @@
+//
+
+use nix::sys;
+use std::{
+    fmt,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread::JoinHandle,
+    time::{Duration, Instant},
+};
+use crate::executor::TaskQueueHandle;
+
+pub struct StallDetection<'a> {
+    executor: usize,
+    queue_handle: TaskQueueHandle,
+    queue_name: &'a str,
+    trace: backtrace::Backtrace,
+    budget: Duration,
+    overage: Duration,
+}
+
+impl fmt::Debug for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("StallDetection")
+            .field("executor", &self.executor)
+            .field("queue_handle", &self.queue_handle)
+            .field("queue_name", &self.queue_name)
+            .field("trace", &self.trace)
+            .field("budget", &self.budget)
+            .field("overage", &self.overage)
+            .finish()
+    }
+}
+
+impl fmt::Display for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[stall-detector -- executor {}] task queue {} went over-budget: {:#?} (budget: \
+             {:#?}). Backtrace: {:#?}",
+            self.executor, self.queue_name, self.overage, self.budget, self.trace,
+        )
+    }
+}
+
+pub trait StallDetectionHandler: std::fmt::Debug + Send + Sync {
+    /// How far past the preemption timer should qualify as a stall
+    /// If None is returned, don't use the stall detector for this task queue.
+    fn high_water_mark(&self, queue_handle: TaskQueueHandle, max_expected_runtime: Duration) -> Option<Duration>;
+
+    /// What signal number to use; see values in libc::SIG*
+    fn signal(&self) -> u8;
+
+    /// Handler called when a task exceeds its budget
+    fn stall(&self, detection: StallDetection<'_>);
+}
+
+#[derive(Debug)]
+pub struct DefaultStallDetectionHandler {}
+
+impl StallDetectionHandler for DefaultStallDetectionHandler {
+    /// The default high water mark is 10% of the preemption time,
+    /// capped at 10ms.
+    fn high_water_mark(&self, _queue_handle: TaskQueueHandle, max_expected_runtime: Duration) -> Option<Duration> {
+        // We consider a queue to be stalling the system if it failed to yield in due
+        // time. For a given maximum expected runtime, we allow a margin of error f 10%
+        // (and an absolute minimum of 10ms) after which we record a stacktrace. i.e. a
+        // task queue has should return shortly after `need_preempt()` returns
+        // true or the stall detector triggers. For example::
+        // * If a task queue has a preempt timer of 100ms the the stall detector
+        // triggers if it doesn't yield after running for 110ms.
+        // * If a task queue has a preempt timer of 5ms the the stall detector
+        // triggers if it doesn't yield after running for 15ms.
+        Some(Duration::from_millis((max_expected_runtime.as_millis() as f64 * 0.1) as u64)
+            .max(Duration::from_millis(10)))
+    }
```

Comment:
This is the impl for `DefaultStallDetectionHandler`?  I thought our customization method was going to be for someone to implement the `StallDetectionHandler` trait. I'm not a fan of knobs on knobs when it comes to configuration. It becomes increasingly difficult to figure out what exactly is going on. If there's a  `DefaultStallDetectionHandler`, we know exactly what values it's running with.

That being said, I'll bow to consensus if you and @glommer want to make it configurable another layer.

Reply:
I don't think calling this the "default" handler is a good idea. Stall detection will be disabled by default and users will have to provide this handler explicitly to enable it. 

I would call it a logging handler, personally since that's what it does. 

As for configuration maybe @glommer has a strong opinion on this. I feel like different applications will have different toleration levels for stalls and so it's not going to be possible for us to give good constants that will work for everyone. 

Instead, I would make it completely customizable and implement `Default`.

Reply:
Sounds good on renaming and adding the customization bits.